nessie/multipart_test.go

```diff
@@ -0,0 +1,115 @@
+// +build systemtests
+
+package nessie
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/stretchr/testify/require"
+	"github.com/thanhpk/randstr"
+	"github.com/treeverse/lakefs/logging"
+)
+
+const (
+	numberOfParts = 5
+	partSize      = 6 * 1024 * 1024
+)
+
+func TestMultiPart(t *testing.T) {
+	ctx, logger, repo := setupTest(t)
+	file := "multipart_file"
+	path := masterBranch + "/" + file
+	input := &s3.CreateMultipartUploadInput{
+		Bucket: aws.String(repo),
+		Key:    aws.String(path),
+	}
+
+	resp, err := svc.CreateMultipartUpload(input)
+	require.NoError(t, err, "failed to create multipart upload")
+	logger.Info("Created multipart upload request")
+
+	completedParts := uploadRandomParts(t, logger, resp)
+
+	completeResponse, err := completeMultipartUpload(svc, resp, completedParts)
+	require.NoError(t, err, "failed to complete multipart upload")
+
+	logger.WithField("key", completeResponse.Key).Info("Completed multipart request successfully")
+
+	var b bytes.Buffer
+	_, err = client.GetObject(ctx, repo, masterBranch, file, &b)
+	require.NoError(t, err, "failed to get object")
+}
+
+func uploadRandomParts(t *testing.T, logger logging.Logger, resp *s3.CreateMultipartUploadOutput) []*s3.CompletedPart {
+	ch := make(chan *s3.CompletedPart, numberOfParts)
+	var wg sync.WaitGroup
+	var anyError sync.Once
+	var uploadErr error
+
+	wg.Add(numberOfParts)
+	for partNumber := 1; partNumber <= numberOfParts; partNumber++ {
+		go func(partNumber int) {
+			completedPart, err := uploadPart(logger, svc, resp, randstr.Bytes(partSize), partNumber)
+
+			if err != nil {
+				// keep just the first error
+				anyError.Do(func() {
```

Comment:
Consider just streaming out all the errors on a channel so we can report them all: when it fails, there's a big difference between one failing and multiple failures.